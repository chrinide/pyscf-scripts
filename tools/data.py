#!/usr/bin/env python

import numpy

nlm = numpy.zeros((56,3), dtype=numpy.int32)

# p's
nlm[1,0] =  1 #px
nlm[2,1] =  1 #py
nlm[3,2] =  1 #pz

# d's
nlm[4,0] =  2  #xx
nlm[5,1] =  2  #yy
nlm[6,2] =  2  #zz
nlm[7,0] =  1  #xy
nlm[7,1] =  1
nlm[8,0] =  1  #xz
nlm[8,2] =  1
nlm[9,1] =  1  #yz
nlm[9,2] =  1

# f's
nlm[10,0] = 3 #xxx
nlm[11,1] = 3 #yyy
nlm[12,2] = 3 #zzz    
nlm[13,0] = 2 #xxy
nlm[13,1] = 1
nlm[14,0] = 2 #xxz
nlm[14,2] = 1
nlm[15,1] = 2 #yyz
nlm[15,2] = 1 
nlm[16,0] = 1 #xyy
nlm[16,1] = 2 
nlm[17,0] = 1 #xzz
nlm[17,2] = 2 
nlm[18,1] = 1 #yzz
nlm[18,2] = 2
nlm[19,0] = 1 #xyz
nlm[19,1] = 1 
nlm[19,2] = 1 

# g's
nlm[20,0] = 4 #xxxx
nlm[21,1] = 4 #yyyy
nlm[22,2] = 4 #zzzz
nlm[23,0] = 3 #xxxy
nlm[23,1] = 1
nlm[24,0] = 3 #xxxz
nlm[24,2] = 1 
nlm[25,0] = 1 #xyyy
nlm[25,1] = 3 
nlm[26,1] = 3 #yyyz
nlm[26,2] = 1 
nlm[27,0] = 1 #xzzz 
nlm[27,2] = 3 
nlm[28,1] = 1 #yzzz
nlm[28,2] = 3
nlm[29,0] = 2 #xxyy
nlm[29,1] = 2 
nlm[30,0] = 2 #xxzz
nlm[30,2] = 2
nlm[31,1] = 2 #yyzz 
nlm[31,2] = 2 
nlm[32,0] = 2 #xxyz 
nlm[32,1] = 1
nlm[32,2] = 1
nlm[33,0] = 1 #xyyz
nlm[33,1] = 2 
nlm[33,2] = 1
nlm[34,0] = 1 #xyzz
nlm[34,1] = 1
nlm[34,2] = 2

