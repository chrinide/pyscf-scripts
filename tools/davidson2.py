#!/usr/bin/env python

from __future__ import division
import math
import numpy as np
import time

''' Block Davidson, Joshua Goings (2013)

    Block Davidson method for finding the first few
        lowest eigenvalues of a large, diagonally dominant,
    sparse Hermitian matrix (e.g. Hamiltonian)
'''

n = 1200                                        # Dimension of matrix
tol = 1e-8                              # Convergence tolerance
mmax = n//2                             # Maximum number of iterations  

''' Create sparse, diagonally dominant matrix A with 
        diagonal containing 1,2,3,...n. The eigenvalues
    should be very close to these values. You can 
    change the sparsity. A smaller number for sparsity
    increases the diagonal dominance. Larger values
    (e.g. sparsity = 1) create a dense matrix
'''

sparsity = 0.0001
A = np.zeros((n,n))
for i in range(0,n):
    A[i,i] = i + 1 
A = A + sparsity*np.random.randn(n,n) 
A = (A.T + A)/2 


k = 8                                   # number of initial guess vectors 
eig = 4                                 # number of eignvalues to solve 
t = np.eye(n,k)                 # set of k unit vectors as guess
V = np.zeros((n,n))             # array of zeros to hold guess vec
I = np.eye(n)                   # identity matrix same dimen as A

# Begin block Davidson routine

start_davidson = time.time()

for m in xrange(k,mmax,k):
    if m <= k:
        for j in xrange(0,k):
            V[:,j] = t[:,j]/np.linalg.norm(t[:,j])
        theta_old = 1 
    elif m > k:
        theta_old = theta[:eig]
    V,R = np.linalg.qr(V)
    T = np.dot(V[:,:(m+1)].T,np.dot(A,V[:,:(m+1)]))
    THETA,S = np.linalg.eig(T)
    idx = THETA.argsort()
    theta = THETA[idx]
    s = S[:,idx]
    for j in xrange(0,k):
        w = np.dot((A - theta[j]*I),np.dot(V[:,:(m+1)],s[:,j])) 
#        q = np.dot(np.linalg.inv(theta[j]*I - np.diag(np.diag(A))),w)
        q = w/(theta[j]-A[j,j])
        V[:,(m+j+1)] = q
    norm = np.linalg.norm(theta[:eig] - theta_old)
    if norm < tol:
        break

end_davidson = time.time()

# End of block Davidson. Print results.

print "davidson = ", theta[:eig],";",\
    end_davidson - start_davidson, "seconds"

# Begin Numpy diagonalization of A

start_numpy = time.time()

E,Vec = np.linalg.eig(A)
E = np.sort(E)

end_numpy = time.time()

# End of Numpy diagonalization. Print results.

print "numpy = ", E[:eig],";",\
     end_numpy - start_numpy, "seconds" 

########################################################

# Build a fake sparse symmetric matrix 
n = 1600
print('Dimension of the matrix',n,'*',n)
sparsity = 0.001
A = np.zeros((n,n))
for i in range(0,n) : 
    A[i,i] = i-9
A = A + sparsity*np.random.randn(n,n)
A = (A.T + A)/2

tol = 1e-9             # Convergence tolerance
mmax = 20              # Maximum number of iterations

# Setup the subspace trial vectors
k = 4
print ('No. of start vectors:',k)
neig = 3
print ('No. of desired Eigenvalues:',neig)
t = np.eye(n,k) # initial trial vectors
v = np.zeros((n,n)) # holder for trial vectors as iterations progress
I = np.eye(n) # n*n identity matrix
ritz = np.zeros((n,n))
f = np.zeros((n,n))
#-------------------------------------------------------------------------------
# Begin iterations  
#-------------------------------------------------------------------------------
start = time.time()
iter = 0
for m in range(k,mmax,k):
    iter = iter + 1
    print ("Iteration no:", iter)
    if iter==1:  # for first iteration add normalized guess vectors to matrix v
        for l in range(m):
            v[:,l] = t[:,l]/(np.linalg.norm(t[:,l]))
    # Matrix-vector products, form the projected Hamiltonian in the subspace
    T = np.linalg.multi_dot([v[:,:m].T,A,v[:,:m]]) # selects fastest evaluation order
    w, vects = np.linalg.eig(T) # Diagonalize the subspace Hamiltonian
    jj = 0
    s = w.argsort()
    ss = w[s]
    #***************************************************************************
    # For each eigenvector of T build a Ritz vector, precondition it and check
    # if the norm is greater than a set threshold.
    #***************************************************************************
    for ii in range(m): #for each new eigenvector of T
        f = np.diag(1./ np.diag((np.diag(np.diag(A)) - w[ii]*I)))
#        print (f)
        ritz[:,ii] = np.dot(f,np.linalg.multi_dot([(A-w[ii]*I),v[:,:m],vects[:,ii]]))
        if np.linalg.norm(ritz[:,ii]) > 1e-7 :
            ritz[:,ii] = ritz[:,ii]/(np.linalg.norm(ritz[:,ii]))
            v[:,m+jj] = ritz[:,ii]
            jj = jj + 1
    q, r = np.linalg.qr(v[:,:m+jj-1])
    for kk in range(m+jj-1):
        v[:,kk] = q[:,kk]
    for ii in range(neig):
        print (ss[ii])
    if iter==1: 
        check_old = ss[:neig]
        check_new = 1
    elif iter==2:
        check_new = ss[:neig]
    else: 
        check_old = check_new
        check_new = ss[:neig]
    check = np.linalg.norm(check_new - check_old)
    if check < tol:
        print('Block Davidson converged at iteration no.:',iter)
        break
end = time.time()
print ('Block Davidson time:',end-start)
start = time.time()
eig, eigvecs = np.linalg.eig(A)
end = time.time() 
s = eig.argsort()
ss = eig[s]
print('Exact Diagonalization:')
for ii in range(neig):    
    print(ss[ii])
#print(ss[:neig])
print ('Exact Diagonalization time:',end-start)
     
