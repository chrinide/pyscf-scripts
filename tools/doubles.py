#!/usr/bin/env python

def doubleexcitations(nocc,nvir):
    doubles = []
    for occa in nocc:
        for occb in nocc:
            for virta in nvir:
                for virtb in nvir:
                    doubles.append((occa,occb,virta,virtb))
    return doubles
