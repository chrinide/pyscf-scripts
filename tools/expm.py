#!/usr/bin/env python

import numpy

def expm(a,tol=1e-6,maxit=40):
    """
    http://epubs.siam.org/doi/abs/10.1137/S00361445024180
    """
    n,m = a.shape
    assert(n==m)
    factor = 1.0
    x =  numpy.identity(n,'d')
    em = numpy.identity(n,'d')
    for j in range(1,maxit):
        factor /= j
        x = numpy.dot(x,a)
        if numpy.linalg.norm(x) < tol:
            break
        em += factor*x
    else:
        print("expm remainder = \n%s" % x)
        raise Exception("Maximum iterations reached in expm")
    return em

