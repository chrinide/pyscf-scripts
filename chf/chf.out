#INFO: **** input file is /home/jluis/src/git/scripts/chf/chf.py ****
#!/usr/bin/env python

import numpy
from functools import reduce
from pyscf import gto, scf, lo, dft, lib
from pyscf.tools import wfn_format, molden

def get_pop(dm, s, lmultipliers):

    label = mol.spheric_labels(False)

    #pop = numpy.einsum('ij,ji->i', dm,s)
    #chg1 = numpy.zeros(mol.natm)
    #qq1 = numpy.zeros(mol.natm)
    #for i, s1 in enumerate(label):
    #    chg1[s1[0]] += pop[i]
    #for ia in range(mol.natm):
    #    symb = mol.atom_symbol(ia)
    #    qq1[ia] = mol.atom_charge(ia)-chg1[ia]
    #    lib.logger.info(mf, 'Pop, Q, of %d %s = %12.6f %12.6f' \
    #    % (ia+1, symb, chg1[ia], qq1[ia]))

    nao = dm.shape[0]
    fock = numpy.zeros_like(dm)
    for ia in range(mol.natm):
        g = numpy.zeros_like(dm)
        iden = numpy.zeros_like(dm)
        for i, s1 in enumerate(label):
            if (s1[0]==ia):
                iden[i,:] = 1.0
        iden = 0.5*(iden+iden.T)
        #print iden
        #tmp = numpy.einsum('ij,ji->ij',iden,s)
        #tmp = numpy.einsum('ij,ji->',tmp,dm)
        #lib.logger.info(mf, 'Pop of %d = %12.6f' % (ia+1, tmp))
        g = numpy.einsum('ij,ji->ij', iden,s)
        fock += lmultipliers[ia]*g

    return fock

def get_cfock(h1e, s1e, vhf, dm, cycle=0, mf_diis=None):
    fock = old_get_fock(h1e, s1e, vhf, dm, cycle, mf_diis)
    fock -= get_pop(dm, s1e, multiplicadores) 
    return fock

name = 'h2'    

mol = gto.Mole()
mol.verbose = 4
mol.atom = '''
H  0.000000000000000  0.000000000000000  0.000000000000000
H  0.000000000000000  0.000000000000000  0.750000000000000
'''
#Li  0.000000000000000  0.000000000000000  0.000000000000000
#F   0.000000000000000  0.000000000000000  1.600000000000000
mol.basis = 'sto-3g'
mol.spin = 0
mol.charge = 0
mol.symmetry = 0
mol.build()

mf = scf.RHF(mol)
mf.verbose = 0
mf.kernel()
dm = mf.make_rdm1()

multiplicadores = numpy.zeros(mol.natm)

my_diis_obj = scf.ADIIS()
my_diis_obj.diis_space = 14

x = [-0.1, -0.2, -0.3, -0.4, -0.5, -0.6, -0.7, -0.8, -0.9, -1.0, 0.0,
0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]
x = numpy.asarray(x)
ehf = []

#for b in x:
#    multiplicadores = [0.0,b]

multiplicadores = [0.0,0.2]
mf = scf.RHF(mol)
mf.conv_tol = 1e-6
mf.max_cycle = 120
mf.diis = my_diis_obj
old_get_fock = mf.get_fock
mf.get_fock = get_cfock 
#mf.verbose = 3
ehf.append(mf.kernel(dm0=dm))
mf.analyze(with_meta_lowdin=False)
dm = mf.make_rdm1()

wfn_file = name + '.wfn'
with open(wfn_file, 'w') as f2:
    wfn_format.write_mo(f2, mol, mf.mo_coeff[:,mf.mo_occ>0], mo_occ=mf.mo_occ[mf.mo_occ>0])
#INFO: ******************** input file end ********************


System: ('Linux', 'silicio', '3.16.0-4-amd64', '#1 SMP Debian 3.16.51-3 (2017-12-13)', 'x86_64', '')  Threads 12
Python 2.7.15 (default, Oct 25 2018, 01:35:55) 
[GCC 8.2.0]
numpy 1.15.3  scipy 1.1.0
Date: Thu Nov  1 02:32:04 2018
PySCF version 1.6a
PySCF path  /home/jluis/src/pyscf/dev/pyscf
GIT ORIG_HEAD 090d7a77795211ecfcdfc6ced606736fde25d1c7
GIT HEAD      ref: refs/heads/dev
GIT dev branch  27eb228cd9fdd7aebc89cf429224ff485f9c21c4

[ENV] PYSCF_TMPDIR /scratch-ssd/jluis
[CONFIG] conf_file None
[INPUT] verbose = 4
[INPUT] num. atoms = 2
[INPUT] num. electrons = 2
[INPUT] charge = 0
[INPUT] spin (= nelec alpha-beta = 2S) = 0
[INPUT] symmetry 0 subgroup None
[INPUT] Mole.unit = angstrom
[INPUT]  1 H      0.000000000000   0.000000000000   0.000000000000 AA    0.000000000000   0.000000000000   0.000000000000 Bohr
[INPUT]  2 H      0.000000000000   0.000000000000   0.750000000000 AA    0.000000000000   0.000000000000   1.417294593424 Bohr

nuclear repulsion = 0.70556961456
number of shells = 2
number of NR pGTOs = 6
number of NR cGTOs = 2
basis = sto-3g
ecp = {}
CPU time:         0.34


******** <class 'pyscf.scf.hf.RHF'> ********
method = RHF
initial guess = minao
damping factor = 0
level shift factor = 0
DIIS = <pyscf.scf.diis.ADIIS object at 0x2b4f3c12a750>
DIIS start cycle = 1
DIIS space = 6
SCF tol = 1e-06
SCF gradient tol = None
max. SCF cycles = 120
direct_scf = True
direct_scf_tol = 1e-13
chkfile to save SCF result = /scratch-ssd/jluis/tmphVcqJf
max_memory 4000 MB (current use 70 MB)
Set gradient conv threshold to 0.001
init E= -1.1161514489386
  HOMO = -0.688402074594307  LUMO = 0.574875567604324
cycle= 1 E= -1.09150077370504  delta_E= 0.0247  |g|= 0.026  |ddm|= 0.396
  HOMO = -0.686010391814613  LUMO = 0.572737786938551
cycle= 2 E= -1.08647818211398  delta_E= 0.00502  |g|= 0.00276  |ddm|= 0.0399
  HOMO = -0.686010391814613  LUMO = 0.572737786938551
cycle= 3 E= -1.08647818211398  delta_E= -4.44e-16  |g|= 0.00276  |ddm|= 7.87e-16
  HOMO = -0.686010391814613  LUMO = 0.572737786938551
cycle= 4 E= -1.08647818211398  delta_E= 4.44e-16  |g|= 0.00276  |ddm|= 7.87e-16
  HOMO = -0.686010391814613  LUMO = 0.572737786938551
cycle= 5 E= -1.08647818211398  delta_E=    0  |g|= 0.00276  |ddm|=    0
  HOMO = -0.686010391814613  LUMO = 0.572737786938551
cycle= 6 E= -1.08647818211398  delta_E=    0  |g|= 0.00276  |ddm|=    0
  HOMO = -0.686010391814613  LUMO = 0.572737786938551
cycle= 7 E= -1.08647818211398  delta_E=    0  |g|= 0.00276  |ddm|=    0
  HOMO = -0.685247248281987  LUMO = 0.572026587815863
cycle= 8 E= -1.08591697440064  delta_E= 0.000561  |g|= 0.000296  |ddm|= 0.00426
  HOMO = -0.685247248281987  LUMO = 0.572026587815863
cycle= 9 E= -1.08591697440064  delta_E= 4.44e-16  |g|= 0.000296  |ddm|= 3.51e-16
  HOMO = -0.685160554621895  LUMO = 0.571945702739841
Extra cycle  E= -1.08585647132366  delta_E= 6.05e-05  |g|= 3.18e-05  |ddm|= 0.000457
converged SCF energy = -1.08585647132366
**** MO energy ****
MO #1   energy= -0.685160554621895 occ= 2
MO #2   energy= 0.571945702739841  occ= 0
 ** Mulliken pop  **
pop of  0 H 1s        0.69406
pop of  1 H 1s        1.30594
 ** Mulliken atomic charges  **
charge of  0H =      0.30594
charge of  1H =     -0.30594
Dipole moment(X, Y, Z, Debye):  0.00000,  0.00000, -1.10213
