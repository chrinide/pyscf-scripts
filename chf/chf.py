#!/usr/bin/env python

import numpy
from functools import reduce
from pyscf import gto, scf, dft, lib
from pyscf.tools import wfn_format

def get_pop(dm, s, lmultipliers):

    label = mol.spheric_labels(False)
    nao = dm.shape[0]
    fock = numpy.zeros_like(dm)
    for ia in range(mol.natm):
        g = numpy.zeros_like(dm)
        iden = numpy.zeros_like(dm)
        for i, s1 in enumerate(label):
            if (s1[0]==ia):
                iden[i,:] = 1.0
        iden = 0.5*(iden+iden.T)
        #print iden
        #tmp = numpy.einsum('ij,ji->ij',iden,s)
        #tmp = numpy.einsum('ij,ji->',tmp,dm)
        #lib.logger.info(mf, 'Pop of %d = %12.6f' % (ia+1, tmp))
        g = numpy.einsum('ij,ji->ij', iden,s)
        fock += lmultipliers[ia]*g

    return fock

def get_cfock(h1e, s1e, vhf, dm, cycle=0, mf_diis=None):
    fock = old_get_fock(h1e, s1e, vhf, dm, cycle, mf_diis)
    fock -= get_pop(dm, s1e, multiplicadores) 
    return fock

name = 'h2'    

mol = gto.Mole()
mol.verbose = 4
mol.atom = '''
H  0.000000000000000  0.000000000000000  0.000000000000000
H  0.000000000000000  0.000000000000000  0.750000000000000
'''
mol.basis = 'sto-3g'
mol.spin = 0
mol.charge = 0
mol.symmetry = 0
mol.build()

mf = scf.RHF(mol)
mf.verbose = 4
mf.kernel()
mf.analyze(with_meta_lowdin=False)

dm = mf.make_rdm1()
multiplicadores = numpy.zeros(mol.natm)
multiplicadores = [0.0,0.2]

my_diis_obj = scf.ADIIS()
my_diis_obj.diis_space = 14
mf = scf.RHF(mol)
mf.conv_tol = 1e-6
mf.max_cycle = 120
mf.diis = my_diis_obj
old_get_fock = mf.get_fock
mf.get_fock = get_cfock 
#mf.verbose = 3
mf.kernel(dm0=dm)
mf.analyze(with_meta_lowdin=False)
dm = mf.make_rdm1()

wfn_file = name + '.wfn'
with open(wfn_file, 'w') as f2:
    wfn_format.write_mo(f2, mol, mf.mo_coeff[:,mf.mo_occ>0], \
    mo_occ=mf.mo_occ[mf.mo_occ>0], mo_energy=mf.mo_energy[mf.mo_occ>0])

