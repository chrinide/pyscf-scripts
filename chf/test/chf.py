#!/usr/bin/env python

import numpy
from functools import reduce
from pyscf import gto, scf, lo, dft, lib
from pyscf.tools import wfn_format, molden

def get_pop(dm, s, lmultipliers):

    label = mol.spheric_labels(False)

    #pop = numpy.einsum('ij,ji->i', dm,s)
    #chg1 = numpy.zeros(mol.natm)
    #qq1 = numpy.zeros(mol.natm)
    #for i, s1 in enumerate(label):
    #    chg1[s1[0]] += pop[i]
    #for ia in range(mol.natm):
    #    symb = mol.atom_symbol(ia)
    #    qq1[ia] = mol.atom_charge(ia)-chg1[ia]
    #    lib.logger.info(mf, 'Pop, Q, of %d %s = %12.6f %12.6f' \
    #    % (ia+1, symb, chg1[ia], qq1[ia]))

    nao = dm.shape[0]
    fock = numpy.zeros_like(dm)
    for ia in range(mol.natm):
        g = numpy.zeros_like(dm)
        iden = numpy.zeros_like(dm)
        for i, s1 in enumerate(label):
            if (s1[0]==ia):
                iden[i,:] = 1.0
        iden = 0.5*(iden+iden.T)
        print iden
        #tmp = numpy.einsum('ij,ji->ij',iden,s)
        #tmp = numpy.einsum('ij,ji->',tmp,dm)
        #lib.logger.info(mf, 'Pop of %d = %12.6f' % (ia+1, tmp))
        g = numpy.einsum('ij,ji->ij', iden,s)
        fock += lmultipliers[ia]*g

    return fock

def get_cfock(h1e, s1e, vhf, dm, cycle=0, mf_diis=None):
    fock = old_get_fock(h1e, s1e, vhf, dm, cycle, mf_diis)
    fock -= get_pop(dm, s1e, multiplicadores) 
    return fock

mol = gto.Mole()
mol.verbose = 4
mol.atom = '''
Li  0.000000000000000  0.000000000000000  0.000000000000000
H   0.000000000000000  0.000000000000000  1.563900000000000
'''
mol.basis = 'sto-3g'
mol.spin = 0
mol.charge = 0
mol.symmetry = 0
mol.build()

mf = scf.RHF(mol)
mf.verbose = 0
mf.kernel()
dm = mf.make_rdm1()

multiplicadores = numpy.zeros(mol.natm)

my_diis_obj = scf.ADIIS()
my_diis_obj.diis_space = 14

x = [-0.1, -0.2, -0.3, -0.4, -0.5, -0.6, -0.7, -0.8, -0.9, -1.0, 0.0,
0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]
x = numpy.asarray(x)
ehf = []

#for b in x:
#    multiplicadores = [0.0,b]

multiplicadores = [0.0,0.1]
mf = scf.RHF(mol)
mf.conv_tol = 1e-6
mf.max_cycle = 120
mf.diis = my_diis_obj
old_get_fock = mf.get_fock
mf.get_fock = get_cfock 
#mf.verbose = 3
ehf.append(mf.kernel(dm0=dm))
mf.analyze(with_meta_lowdin=False)
dm = mf.make_rdm1()

#name = 'lif_0p6'    
#wfn_file = name + '.wfn'
#with open(wfn_file, 'w') as f2:
#    wfn_format.write_mo(f2, mol, mf.mo_coeff[:,mf.mo_occ>0], \
#    mo_occ=mf.mo_occ[mf.mo_occ>0], mo_energy=mf.mo_energy[mf.mo_occ>0])

