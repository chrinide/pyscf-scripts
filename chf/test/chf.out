#INFO: **** input file is /home/jluis/src/git/scripts/chf/test/chf.py ****
#!/usr/bin/env python

import numpy
from functools import reduce
from pyscf import gto, scf, lo, dft, lib
from pyscf.tools import wfn_format, molden

def get_pop(dm, s, lmultipliers):

    label = mol.spheric_labels(False)

    #pop = numpy.einsum('ij,ji->i', dm,s)
    #chg1 = numpy.zeros(mol.natm)
    #qq1 = numpy.zeros(mol.natm)
    #for i, s1 in enumerate(label):
    #    chg1[s1[0]] += pop[i]
    #for ia in range(mol.natm):
    #    symb = mol.atom_symbol(ia)
    #    qq1[ia] = mol.atom_charge(ia)-chg1[ia]
    #    lib.logger.info(mf, 'Pop, Q, of %d %s = %12.6f %12.6f' \
    #    % (ia+1, symb, chg1[ia], qq1[ia]))

    nao = dm.shape[0]
    fock = numpy.zeros_like(dm)
    for ia in range(mol.natm):
        g = numpy.zeros_like(dm)
        iden = numpy.zeros_like(dm)
        for i, s1 in enumerate(label):
            if (s1[0]==ia):
                iden[i,:] = 1.0
        iden = 0.5*(iden+iden.T)
        #print iden
        #tmp = numpy.einsum('ij,ji->ij',iden,s)
        #tmp = numpy.einsum('ij,ji->',tmp,dm)
        #lib.logger.info(mf, 'Pop of %d = %12.6f' % (ia+1, tmp))
        g = numpy.einsum('ij,ji->ij', iden,s)
        fock += lmultipliers[ia]*g

    return fock

def get_cfock(h1e, s1e, vhf, dm, cycle=0, mf_diis=None):
    fock = old_get_fock(h1e, s1e, vhf, dm, cycle, mf_diis)
    fock -= get_pop(dm, s1e, multiplicadores) 
    return fock

mol = gto.Mole()
mol.verbose = 4
mol.atom = '''
H  0.000000000000000  0.000000000000000  0.000000000000000
H  0.000000000000000  0.000000000000000  0.750000000000000
'''
#Li  0.000000000000000  0.000000000000000  0.000000000000000
#F   0.000000000000000  0.000000000000000  1.600000000000000
mol.basis = 'sto-3g'
mol.spin = 0
mol.charge = 0
mol.symmetry = 0
mol.build()

mf = scf.RHF(mol)
mf.verbose = 0
mf.kernel()
dm = mf.make_rdm1()

multiplicadores = numpy.zeros(mol.natm)

my_diis_obj = scf.ADIIS()
my_diis_obj.diis_space = 14

x = [-0.1, -0.2, -0.3, -0.4, -0.5, -0.6, -0.7, -0.8, -0.9, -1.0, 0.0,
0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]
x = numpy.asarray(x)
ehf = []

#for b in x:
#    multiplicadores = [0.0,b]

multiplicadores = [0.0,0.1]
mf = scf.RHF(mol)
mf.conv_tol = 1e-6
mf.max_cycle = 120
mf.diis = my_diis_obj
old_get_fock = mf.get_fock
mf.get_fock = get_cfock 
#mf.verbose = 3
ehf.append(mf.kernel(dm0=dm))
mf.analyze(with_meta_lowdin=False)
dm = mf.make_rdm1()

name = 'h2_0p1'    
wfn_file = name + '.wfn'
with open(wfn_file, 'w') as f2:
    wfn_format.write_mo(f2, mol, mf.mo_coeff[:,mf.mo_occ>0], \
    mo_occ=mf.mo_occ[mf.mo_occ>0], mo_energy=mf.mo_energy[mf.mo_occ>0])

#INFO: ******************** input file end ********************


System: ('Linux', 'silicio', '3.16.0-4-amd64', '#1 SMP Debian 3.16.51-3 (2017-12-13)', 'x86_64', '')  Threads 12
Python 2.7.15 (default, Oct 25 2018, 01:35:55) 
[GCC 8.2.0]
numpy 1.15.3  scipy 1.1.0
Date: Thu Nov  1 03:03:54 2018
PySCF version 1.6a
PySCF path  /home/jluis/src/pyscf/dev/pyscf
GIT ORIG_HEAD 090d7a77795211ecfcdfc6ced606736fde25d1c7
GIT HEAD      ref: refs/heads/dev
GIT dev branch  27eb228cd9fdd7aebc89cf429224ff485f9c21c4

[ENV] PYSCF_TMPDIR /scratch-ssd/jluis
[CONFIG] conf_file None
[INPUT] verbose = 4
[INPUT] num. atoms = 2
[INPUT] num. electrons = 2
[INPUT] charge = 0
[INPUT] spin (= nelec alpha-beta = 2S) = 0
[INPUT] symmetry 0 subgroup None
[INPUT] Mole.unit = angstrom
[INPUT]  1 H      0.000000000000   0.000000000000   0.000000000000 AA    0.000000000000   0.000000000000   0.000000000000 Bohr
[INPUT]  2 H      0.000000000000   0.000000000000   0.750000000000 AA    0.000000000000   0.000000000000   1.417294593424 Bohr

nuclear repulsion = 0.70556961456
number of shells = 2
number of NR pGTOs = 6
number of NR cGTOs = 2
basis = sto-3g
ecp = {}
CPU time:         0.35


******** <class 'pyscf.scf.hf.RHF'> ********
method = RHF
initial guess = minao
damping factor = 0
level shift factor = 0
DIIS = <pyscf.scf.diis.ADIIS object at 0x2b13f2662710>
DIIS start cycle = 1
DIIS space = 6
SCF tol = 1e-06
SCF gradient tol = None
max. SCF cycles = 120
direct_scf = True
direct_scf_tol = 1e-13
chkfile to save SCF result = /scratch-ssd/jluis/tmpOhnSNX
max_memory 4000 MB (current use 70 MB)
Set gradient conv threshold to 0.001
init E= -1.1161514489386
  HOMO = -0.627957372638366  LUMO = 0.614430865648383
cycle= 1 E= -1.10981643357851  delta_E= 0.00634  |g|= 0.0126  |ddm|= 0.199
  HOMO = -0.627326046423829  LUMO = 0.61386462638463
cycle= 2 E= -1.10855319953738  delta_E= 0.00126  |g|= 0.00123  |ddm|= 0.0192
  HOMO = -0.627326046423829  LUMO = 0.61386462638463
cycle= 3 E= -1.10855319953738  delta_E=    0  |g|= 0.00123  |ddm|=    0
  HOMO = -0.627326046423829  LUMO = 0.61386462638463
cycle= 4 E= -1.10855319953738  delta_E=    0  |g|= 0.00123  |ddm|=    0
  HOMO = -0.627326046423829  LUMO = 0.61386462638463
cycle= 5 E= -1.10855319953738  delta_E=    0  |g|= 0.00123  |ddm|= 2.29e-16
  HOMO = -0.627326046423829  LUMO = 0.61386462638463
cycle= 6 E= -1.10855319953738  delta_E=    0  |g|= 0.00123  |ddm|= 2.29e-16
  HOMO = -0.627326046423829  LUMO = 0.613864626384629
cycle= 7 E= -1.10855319953738  delta_E=    0  |g|= 0.00123  |ddm|= 2.12e-15
  HOMO = -0.627131172879383  LUMO = 0.613682744989224
cycle= 8 E= -1.10842382099661  delta_E= 0.000129  |g|= 0.00012  |ddm|= 0.00187
  HOMO = -0.627131172879383  LUMO = 0.613682744989225
cycle= 9 E= -1.10842382099661  delta_E=    0  |g|= 0.00012  |ddm|=    0
  HOMO = -0.627110904723702  LUMO = 0.613663807723632
Extra cycle  E= -1.10841110979128  delta_E= 1.27e-05  |g|= 1.18e-05  |ddm|= 0.000183
converged SCF energy = -1.10841110979128
**** MO energy ****
MO #1   energy= -0.627110904723702 occ= 2
MO #2   energy= 0.613663807723632  occ= 0
 ** Mulliken pop  **
pop of  0 H 1s        0.84480
pop of  1 H 1s        1.15520
 ** Mulliken atomic charges  **
charge of  0H =      0.15520
charge of  1H =     -0.15520
Dipole moment(X, Y, Z, Debye):  0.00000,  0.00000, -0.55908
